src/dbnode/storage/shard.go

```diff
@@ -2156,18 +2169,36 @@ func (s *dbShard) removeAnyFlushStatesTooEarly(tickStart time.Time) {
 
 func (s *dbShard) CleanupExpiredFileSets(earliestToRetain time.Time) error {
 	filePathPrefix := s.opts.CommitLogOptions().FilesystemOptions().FilePathPrefix()
-	multiErr := xerrors.NewMultiError()
-	expired, err := s.filesetBeforeFn(filePathPrefix, s.namespace.ID(), s.ID(), earliestToRetain)
+	expired, err := s.filesetPathsBeforeFn(filePathPrefix, s.namespace.ID(), s.ID(), earliestToRetain)
 	if err != nil {
-		detailedErr :=
-			fmt.Errorf("encountered errors when getting fileset files for prefix %s namespace %s shard %d: %v",
-				filePathPrefix, s.namespace.ID(), s.ID(), err)
-		multiErr = multiErr.Add(detailedErr)
+		return fmt.Errorf("encountered errors when getting fileset files for prefix %s namespace %s shard %d: %v",
+			filePathPrefix, s.namespace.ID(), s.ID(), err)
 	}
-	if err := s.deleteFilesFn(expired); err != nil {
-		multiErr = multiErr.Add(err)
+
+	return s.deleteFilesFn(expired)
+}
+
+func (s *dbShard) CleanupCompactedFileSets() error {
+	filePathPrefix := s.opts.CommitLogOptions().FilesystemOptions().FilePathPrefix()
+	filesets, err := s.filesetsFn(filePathPrefix, s.namespace.ID(), s.ID())
+	if err != nil {
+		return fmt.Errorf("encountered errors when getting fileset files for prefix %s namespace %s shard %d: %v",
+			filePathPrefix, s.namespace.ID(), s.ID(), err)
 	}
-	return multiErr.FinalError()
+	// Get a snapshot of all states here to prevent constantly getting/releasing
+	// locks in a tight loop below.
+	blockStates := s.BlockStatesSnapshot()
+	// Filter without allocating by using same backing array.
+	compacted := filesets[:0]
+	for _, datafile := range filesets {
```

Comment:
maybe add a comment saying this is safe because cleanup and flushing never happen in parallel so the snapshot will never get stale.

Also do we have `BlockStateSnapshot()` and `FlushState()` do those two duplicate each other?

Reply:
Sure I'll add a comment. Also, since versions are always increasing, even if we get stale snapshots, we just don't clean up some files that can be cleaned (but we'll clean them up eventually next time round).

`BlockStatesSnapshot` gets the flush states for all block starts, while `FlushState` is for a specific block start. Wouldn't want to loop through all blocks and keep acquiring and releasing locks (mentioned in a comment above).